docs/02_cleos/03_command-reference/push/push-action.md

```diff
@@ -31,3 +29,65 @@ Push a transaction with a single action
 `--max-net-usage` _UINT_ - set an upper limit on the net usage budget, in bytes, for the transaction (defaults to 0 which means no limit)
 
 `--delay-sec` _UINT_ - set the delay_sec seconds, defaults to 0s
+
+## Output
+
+For actions that return a value `cleos` will have an extra line which will be prefixed with `=>`.
+
+Below you can see three examples of pushing actions `rstring` and `ruint` which return the value passed as input parameter to them:
+
+```shell
+> cleos push action eosio rstring '{"str":null}'  -p eosio@active
+
+executed transaction: 1e72a3ef1fa01a9c90b870fe86bf31413c6a2f40a2722ca72d9dd707f58851af  96 bytes  116 us
+#         eosio <= eosio::rstring               {"str":""}
+=>                                return value: ""
+
+> cleos push action eosio rstring '{"str":"test"}'  -p eosio@active
+
+executed transaction: 2484ba021683bb3c6daaabe291ba291e2689d1be899a293d0a09cc68cf0967fb  96 bytes  116 us
+#         eosio <= eosio::rstring               {"str":"test"}
+=>                                return value: "test"
+
+> cleos push action eosio ruint '{"i":42}'  -p eosio@active
+
+executed transaction: 12c51fd27fad9bb0fa959037a72093ad0b168f5846e916a0e9a295c8416bdf9f  96 bytes  138 us
+#         eosio <= eosio::ruint                 {"i":42}
+=>                                return value: 42
+```
+
+It can also work with C++ structures or classes as well. For a C++ struct like this:
+
+```cpp
+struct tmp {
+    uint32_t id = 1;
+    std::vector<uint8_t> list{ 1, 2 };
+
+    EOSLIB_SERIALIZE(tmp, (id)(list))
```

Comment:
I would empathize the fact that EOSLIB_SERIALIZE() is a must.

Reply:
I've asked Larry and he told me it is not necessary, how sure are you?

Reply:
inviting @larryk85 to the thread

Reply:
EOSLIB_SERIALIZE is not a must.  It helps in terms of compile time for datastream operations and is a must for "some" user defined types that PFR twirls into infinite on.  We will punt back to Boost::PFR if you don't add that macro and we will replace that with something even better before 2.2, so it's not a requirement for most data types.  We mainly used it for the in library stuff to save on compile time for the end user, but for their contracts the small amount of data types they have is negligible.

Reply:
@larryk85 
how can the users know when they fall into that case of "some" user defined type?
you are mentioning an infinite loop?... will the execution of the action using this specific user defined type brake?
I still think we should mention this case, define it, so if anyone hits it they will know how to get around it: explicitly use the EOSLIB_SERIALIZE

Reply:
I would rather not mention it.  It is a macro that will be going away very soon.  The macro doesn't help the cases that I was thinking about anyway.  PFR will fail on inherited types or types that aren't trivially constructible, which means they would have to define the datastream operators by hand anyway.

Reply:
ok so I will not mention the macro, I will even take it out from the example code in this doc.

but, I realize now that saying "any user defined type can be returned from an action" is not actually accurate. I'd like to make this statement more accurate, something like this: you can return almost any user defined type, however there are some exceptions, those which are not trivially constructible, for which you'll have to define custom the datastream operators to be able to serialize them correctly.

When will the user know they are in trouble with the user defined type they want to return which will need custom datastream operators? The serialization will fail? Will be any error? Will the return value just be empty, not present?

Reply:
But, that restriction has nothing to do with returning a value.  That is a general restriction of datastream.  It will fail at compile time.  We have or should have all of this documented with the datastream system.

Reply:
I took out the macro from this doc page.
the rest of the discussion pertains to the return values from action explainer page, so I will continue it there.